File: sdks/java/testing/nexmark/src/main/java/org/apache/beam/sdk/nexmark/model/AuctionBid.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.beam.sdk.nexmark.model;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.core.JsonProcessingException;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.io.Serializable;
import org.apache.beam.sdk.coders.Coder;
import org.apache.beam.sdk.coders.CoderException;
import org.apache.beam.sdk.coders.CustomCoder;
import org.apache.beam.sdk.nexmark.NexmarkUtils;
import org.apache.beam.sdk.nexmark.queries.WinningBids;
import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.base.Objects;
import org.checkerframework.checker.nullness.qual.Nullable;

/** Result of {@link WinningBids} transform. */
public class AuctionBid implements KnownSize, Serializable {
  public static final Coder<AuctionBid> CODER =
      new CustomCoder<AuctionBid>() {
        @Override
        public void encode(AuctionBid value, OutputStream outStream)
            throws CoderException, IOException {
          Auction.CODER.encode(value.auction, outStream);
          Bid.CODER.encode(value.bid, outStream);
        }

        @Override
        public AuctionBid decode(InputStream inStream) throws CoderException, IOException {
          Auction auction = Auction.CODER.decode(inStream);
          Bid bid = Bid.CODER.decode(inStream);
          return new AuctionBid(auction, bid);
        }

        @Override
        public Object structuralValue(AuctionBid v) {
          return v;
        }
      };

  @JsonProperty public Auction auction;

  @JsonProperty public Bid bid;

  @SuppressWarnings("unused")
  public AuctionBid() {
    auction = null;
    bid = null;
  }

  public AuctionBid(Auction auction, Bid bid) {
    this.auction = auction;
    this.bid = bid;
  }

  @Override
  public long sizeInBytes() {
    return auction.sizeInBytes() + bid.sizeInBytes();
  }

  @Override
  public String toString() {
    try {
      return NexmarkUtils.MAPPER.writeValueAsString(this);
    } catch (JsonProcessingException e) {
      throw new RuntimeException(e);
    }
  }

  @Override
  public boolean equals(@Nullable Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    AuctionBid that = (AuctionBid) o;
    return Objects.equal(auction, that.auction) && Objects.equal(bid, that.bid);
  }

  @Override
  public int hashCode() {
    return Objects.hashCode(auction, bid);
  }
}

File: sdks/python/apache_beam/testing/benchmarks/nexmark/models/auction_bid.py
#
# Licensed to the Apache Software Foundation (ASF) under one or more
# contributor license agreements.  See the NOTICE file distributed with
# this work for additional information regarding copyright ownership.
# The ASF licenses this file to You under the Apache License, Version 2.0
# (the "License"); you may not use this file except in compliance with
# the License.  You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

"""Result of WinningBid transform."""
from apache_beam.coders import coder_impl
from apache_beam.coders.coders import FastCoder
from apache_beam.testing.benchmarks.nexmark import nexmark_util
from apache_beam.testing.benchmarks.nexmark.models import nexmark_model


class AuctionBidCoder(FastCoder):
  def to_type_hint(self):
    return AuctionBid

  def _create_impl(self):
    return AuctionBidCoderImpl()

  def is_deterministic(self):
    return True


class AuctionBid(object):
  CODER = AuctionBidCoder()

  def __init__(self, auction, bid):
    self.auction = auction
    self.bid = bid

  def __repr__(self):
    return nexmark_util.model_to_json(self)


class AuctionBidCoderImpl(coder_impl.StreamCoderImpl):
  _auction_coder_impl = nexmark_model.AuctionCoderImpl()
  _bid_coder_Impl = nexmark_model.BidCoderImpl()

  def encode_to_stream(self, value, stream, nested):
    self._auction_coder_impl.encode_to_stream(value.auction, stream, True)
    self._bid_coder_Impl.encode_to_stream(value.bid, stream, True)

  def decode_from_stream(self, stream, nested):
    auction = self._auction_coder_impl.decode_from_stream(stream, True)
    bid = self._bid_coder_Impl.decode_from_stream(stream, True)
    return AuctionBid(auction, bid)
